### ISSUE TITLE
addBowerPackageToProject is not installing directly from a Github repo

### ISSUE THREAD

**Issue by givanse** [ember-cli/ember-cli#3409]
The blueprint looks like this:

    afterInstall: function(options) {
        return this.addBowerPackageToProject('pagedown',                             
          'git@github.com:ujifgc/pagedown.git#f1ae0f5626');
    }

After doing `npm link` when I `ember g <addon name>` I see this:

    not-cached git://github.com/baminteractive/PageDown.git#git@github.com:ujifgc/pagedown.git#f1ae0f5626

So the issue is that **baminteractive/PageDown** is registered as a bower package and that is being fetched. 

What I want is to install pagedown from another repo, the equivalent to:

    bower install --save git@github.com:ujifgc/pagedown.git#f1ae0f5626

See: https://github.com/ember-cli/ember-cli/blob/master/lib/models/blueprint.js#L846

**Comment by givanse** [comment]
So, how about we handle the case where a name is not given?

Example:
    afterInstall: function(options) {
        return this.addBowerPackageToProject('',                             
          'git@github.com:ujifgc/pagedown.git#f1ae0f5626');
    }

Currently it does this:
    Invalid endpoint: #git@github.com:ujifgc/pagedown.git#f1ae0f5626

So, if a name is not given the hash wouldn't be prepended.

**Comment by leandrocp** [comment]
Considering that bower accepts this kind of name (http://bower.io/docs/api/#install), I think that `addBowerPackageToProject` should accept it too. I can look into it if is desirable.

@rwjblue thoughts ?

**Comment by leandrocp** [comment]
BTW, i´ve created this test https://github.com/leandrocp/ember-cli/commit/3eb9cb750dc9bc49bf13b43a5c8fc441c369a196 to check different bower endpoints

@rwjblue it it a good idea to merge this test ?

**Comment by twokul** [comment]
```javascript
```

**Comment by leandrocp** [comment]
And how this should be handled? 

```
this.addBowerPackageToProject('git@github.com:ujifgc/pagedown.git#f1ae0f5626',  '#commit');
```

See that target commit is different from packageName. 

And that feature should be extended to install npm package too.

**Comment by twokul** [comment]
my brain doesn't work well today so correct me if I'm wrong. `f1ae0f5626` commit sha is included in the url and then we can extract the name of the repo? why do we need to pass the second param?

**Comment by givanse** [comment]
@twokul I like that, passing a single argument would have the effect of:

    bower install --save <single arg>

I don't see why we would pass a second arg. The assumption would be that a single arg has everything (branch/tag). And the current behavior will be followed only if two args are passed.

**Comment by leandrocp** [comment]
@twokul that´s the result:

```
this.addBowerPackageToProject('git@github.com:ujifgc/pagedown.git#f1ae0f5626')
// package: 'git@github.com:ujifgc/pagedown.git#f1ae0f5626'

this.addBowerPackageToProject('git@github.com:ujifgc/pagedown.git#f1ae0f5626', '1.0)
// package: 'git@github.com:ujifgc/pagedown.git#f1ae0f5626#1.0'
```

The point here is that `addBowerPackagesToProject` just send an array of endpoints to `bower.commands.install` (https://github.com/ember-cli/ember-cli/blob/master/lib%2Ftasks%2Fbower-install.js#L28)

**Comment by stefanpenner** [comment]
we are planning to drop bower, so this will likely not be fixed. If someone submits a PR it will be accepted.

**Comment by givanse** [comment]
Is there an issue or branch for this transition?

**Comment by stefanpenner** [comment]
@givanse https://github.com/ember-cli/ember-cli/issues/4211

**Comment by stefanpenner** [comment]
@olivierlesnicki we ultimately just use bowers own code to do this.

@mike-north thoughts?

**Comment by olivierlesnicki** [comment]
@stefanpenner using bower I'm able to list `"repo": "olivierlesnicki/repo#v1.0.0"` as a dependency. Yet I'm unable to leverage addBowerPackagesToProject to carry the same task.

```
this.addBowerPackagesToProject([{
 name: 'repo',
 target: 'olivierlesnicki/repo#v1.0.0'
}])
```

A "can't find package" error gets thrown.

**Comment by mike-north** [comment]
Using [bower's own code](https://github.com/bower/endpoint-parser):

```js
var endpointParser = require('bower-endpoint-parser');
endpointParser.json2decomposed('pagedown', 'git@github.com:ujifgc/pagedown.git#f1ae0f5626');
```
**returns**
```js
{ name: 'pagedown',
  source: 'git@github.com:ujifgc/pagedown.git',
  target: 'f1ae0f5626' }
```
which is what I'd expect

@givanse - which version of ember-cli are you using to report this behavior.

**Comment by givanse** [comment]
@mike-north I was using `0.2.0-beta.1`.

By reviewing the commit history of my addon I found that later I upgraded to `ember-cli/ember-cli#0a21da147d9ac46e879e20b0cec05df673c155ab`. I don't think it changes anything, but I thought it would be best to mention it.

**Comment by mike-north** [comment]
It should be fixed in relatively recent versions of ember-cli. The PR that corrected the strange behavior you've noticed is #4430

**Comment by olivierlesnicki** [comment]
@mike-north confirmed it is fixed
